### ISSUE TITLE
update layers_pattern logic to work independently on multiple modules (i.e. encoder and decoder)

### ISSUE THREAD

**Issue by fcakyon** [huggingface/peft#2165]
### Feature request

update layers_pattern logic so that it is possible to specify layer indices from more than one module (i.e. last 2 layers of encoder and all layers of decoder)

### Motivation

For my PhD research, I need to apply LORA to the last 2 encoder layers and all decoder layers. I can select encoder and decoder layers as below, but the code below matches all encoder and decoder layers.

```python
target_modules = ["q_proj", "k_proj", "v_proj", "out_proj"]

config = LoraConfig(
    target_modules=target_modules,
    layers_pattern=['language_model.model.encoder', 'language_model.model.decoder'],
    layers_to_transform=[0, 1, 2, 3, 4, 5],
    task_type="CAUSAL_LM"
)
```

Then I try to provide all desired layer patterns manually, as below, and nothing matches, raising an error since the current left.BaseTuner.inject_adapter does not allow including the layer index in the pattern.

```python
target_modules = ["q_proj", "k_proj", "v_proj", "out_proj"]

config = LoraConfig(
    target_modules=target_modules,
    layers_pattern=['language_model.model.encoder.layers.4', 'language_model.model.encoder.layers.5', 'language_model.model.decoder.layers.5'],
    task_type="CAUSAL_LM"
)
```

It would be great to support my use case (I believe its not a unique use pattern)

### Your contribution

I am open to work any PR

**Comment by BenjaminBossan** [comment]
When you use `layers_pattern`, you should also use `layers_to_transform` to indicate the indices. I'm not quite sure what model you're using, so I can't say if this will work, but could you try:

```python
layers_pattern=['encoder.layers'], layers_to_transform=[4, 5]
```

As it's not possible to have different groups of patterns and indices, you would have to add the layers you want to target via `'language_model.model.decoder.layers.5'` directly to `target_modules`. You could also check if you have some luck with the newly added `exclude_modules` argument, but that requires installing PEFT from source.

Note that with `layers_pattern` and `layers_to_transform`, we support only a rather simple matching pattern. If you need something very specific, I would recommend to craft a regex and pass that to `target_modules`. That way, you should be able to cover all use cases.

**Comment by fcakyon** [comment]
> Note that with `layers_pattern` and `layers_to_transform`, we support only a rather simple matching pattern. If you need something very specific, I would recommend to craft a regex and pass that to `target_modules`. That way, you should be able to cover all use cases.

Thank you for the response @BenjaminBossan  🙏🏻  Let me paraphrase my point:

Providing custom regexp in `target_modules` doesn't work since this restrictive hardcoded regexp on the peft.BaseTuner.inject_adapter doesn't allow the layer index to be included in the pattern: https://github.com/huggingface/peft/blob/e8259ff7bc72365ec730efb1f4005a95071d5dfd/src/peft/tuners/tuners_utils.py#L986

It's not possible to specify layer index in `target_modules` as `'language_model.model.decoder.layers.5'`. It won't match anything.

Do you have any idea on how to fix this problem?

**Comment by BenjaminBossan** [comment]
> this restrictive hardcoded regexp on the peft.BaseTuner.inject_adapter doesn't allow the layer index to be included in the pattern

Yes, this is on purpose, the index should be specified in `layers_to_transform`, not `layers_pattern`. However, as you mentioned, this won't work for use cases as yours. I would rather not complicate this mechanism further, however, as it can already be quite confusing as is.

> It's not possible to specify layer index in `target_modules` as `'language_model.model.decoder.layers.5'`. It won't match anything.

What I mean is that the whole `target_modules` can be a string, in which case it is interpreted as a regex, instead of passing a list of strings. When you pass a regex, you have basically full control over which layers to match. It can be a bit cumbersome to assemble this regex to exactly fit the layers you want, but it should always be possible. Using an LLM can be quite helpful to quickly create the regex.

**Comment by BenjaminBossan** [comment]
No, when you pass `target_modules="some-regex"`, we never reach that line, see this line:

https://github.com/huggingface/peft/blob/e8259ff7bc72365ec730efb1f4005a95071d5dfd/src/peft/tuners/tuners_utils.py#L963-L964

We will never reach the `else` condition of the line you cited.

**Comment by fcakyon** [comment]
> No, when you pass `target_modules="some-regex"`, we never reach that line, see this line:
> 
> https://github.com/huggingface/peft/blob/e8259ff7bc72365ec730efb1f4005a95071d5dfd/src/peft/tuners/tuners_utils.py#L963-L964
> 
> We will never reach the `else` condition of the line you cited.

I provided 2 regexp strings (List[str]), giving the unwanted result. I combined them into a single regexp string, and now it works!

Thank you for your patience and understanding! 🙏🏻

**Comment by BenjaminBossan** [comment]
I'm glad it worked out, good luck.
